include/sparrow/array_data_factory.hpp

```diff
@@ -0,0 +1,359 @@
+    array_data make_array_data_for_variable_size_binary_layout()
+    {
+        return {
+            .type = data_descriptor(arrow_type_id<T>()),
+            .length = 0,
+            .offset = 0,
+            .bitmap = {},
+            .buffers = {{}, array_data::buffer_type(sizeof(std::int64_t), 0)},
+            .child_data = {},
+            .dictionary = nullptr
+        };
+    }
+
+    template <mpl::constant_range ValueRange>
+        requires std::ranges::range<std::unwrap_ref_decay_t<std::ranges::range_value_t<ValueRange>>>
+    array_data make_array_data_for_variable_size_binary_layout(
+        ValueRange&& values,
+        const array_data::bitmap_type& bitmap,
+        std::int64_t offset
+    )
+    {
+        using U = std::unwrap_ref_decay_t<std::ranges::range_value_t<ValueRange>>;
+        if constexpr (std::ranges::range<U>)
+        {
+            using V = std::ranges::range_value_t<U>;
+            if constexpr (!is_arrow_base_type_extended<V>)
+            {
+                static_assert(false, "Unsupported range value type");
+            }
+        }
+        SPARROW_ASSERT_TRUE(values.size() == bitmap.size());
+        SPARROW_ASSERT_TRUE(std::cmp_greater_equal(values.size(), offset));
+
+        const auto create_buffers = [&values]()
+        {
+            using T = std::ranges::range_value_t<ValueRange>;
+            const auto& unwrap_value = [](const T& value)
+            {
+                if constexpr (mpl::is_reference_wrapper_v<T>)
+                {
+                    return value.get();
+                }
+                else
+                {
+                    return value;
+                }
+            };
+
+            std::vector<array_data::buffer_type> buffers(2);
+            buffers[0].resize(sizeof(std::int64_t) * (values.size() + 1), 0);
+            buffers[1].resize(std::accumulate(
+                values.begin(),
+                values.end(),
+                size_t(0),
+                [&unwrap_value](std::size_t res, const auto& s)
+                {
+                    return res + unwrap_value(s).size();
+                }
+            ));
+            auto iter = buffers[1].begin();
+            const auto offsets = buffers[0].data<std::int64_t>();
+
+            size_t i = 0;
+            for (const auto& value : values)
+            {
+                const auto& unwraped_value = unwrap_value(value);
+                SPARROW_ASSERT_TRUE(
+                    std::cmp_less(unwraped_value.size(), std::numeric_limits<std::int64_t>::max())
+                );
+                offsets[i + 1] = offsets[i] + static_cast<std::int64_t>(unwraped_value.size());
+                std::ranges::copy(unwraped_value, iter);
+                std::advance(iter, unwraped_value.size());
+                ++i;
+            }
+            return buffers;
+        };
+
+        using T = std::unwrap_ref_decay_t<std::unwrap_ref_decay_t<std::ranges::range_value_t<ValueRange>>>;
```

Comment:
Why unwrapping twice?

Reply:
My type is: std::vector<std::reference_wrapper<const std::string>>
The first std::unwrap_ref_decay_t "removes" the reference_wrapper and we get a const std::string&
The second one remove the const and reference and we get the std::string